src/MongoDB.Driver/Encryption/AutoEncryptionOptions.cs

```diff
@@ -0,0 +1,229 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using MongoDB.Bson;
+using MongoDB.Driver.Core.Misc;
+using MongoDB.Shared;
+
+namespace MongoDB.Driver.Encryption
+{
+    /// <summary>
+    /// Auto encryption options.
+    /// </summary>
+    public class AutoEncryptionOptions
+    {
+        // private fields
+        private readonly bool _bypassAutoEncryption;
+        private readonly IReadOnlyDictionary<string, object> _extraOptions;
+        private readonly IMongoClient _keyVaultClient;
+        private readonly CollectionNamespace _keyVaultNamespace;
+        private readonly IReadOnlyDictionary<string, IReadOnlyDictionary<string, object>> _kmsProviders;
+        private readonly IReadOnlyDictionary<string, BsonDocument> _schemaMap;
+
+        // constructors
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AutoEncryptionOptions"/> class.
+        /// </summary>
+        /// <param name="keyVaultNamespace">The keyVault namespace.</param>
+        /// <param name="kmsProviders">The kms providers.</param>
+        /// <param name="bypassAutoEncryption">The bypass auto encryption flag.</param>
+        /// <param name="extraOptions">The extra options.</param>
+        /// <param name="keyVaultClient">The keyVault client.</param>
+        /// <param name="schemaMap">The schema map.</param>
+        public AutoEncryptionOptions(
+            CollectionNamespace keyVaultNamespace,
+            IReadOnlyDictionary<string, IReadOnlyDictionary<string, object>> kmsProviders,
+            Optional<bool> bypassAutoEncryption = default,
+            Optional<IReadOnlyDictionary<string, object>> extraOptions = default,
+            Optional<IMongoClient> keyVaultClient = default,
+            Optional<IReadOnlyDictionary<string, BsonDocument>> schemaMap = default) 
+        {
+            _keyVaultNamespace = Ensure.IsNotNull(keyVaultNamespace, nameof(keyVaultNamespace));
+            _kmsProviders = Ensure.IsNotNull(kmsProviders, nameof(kmsProviders));
+            _bypassAutoEncryption = bypassAutoEncryption.WithDefault(false);
+            _extraOptions = extraOptions.WithDefault(null);
+            _keyVaultClient = keyVaultClient.WithDefault(null);
+            _schemaMap = schemaMap.WithDefault(null);
+
+            EncryptionExtraOptionsValidator.EnsureThatExtraOptionsAreValid(_extraOptions);
```

Comment:
`EncryptionExtraOptionsValidator` and `KmsProvidersHelper` can be joined in a new class(with 2 methods for now) and with a name similar to `EncryptionFieldsValidator`.